lib/codecs/src/decoding/format/vrl.rs

```diff
@@ -0,0 +1,257 @@
+    #[serde(default)]
+    #[configurable(metadata(docs::advanced))]
+    pub timezone: Option<TimeZone>,
+}
+
+impl VrlDeserializerConfig {
+    /// Build the `VrlDeserializer` from this configuration.
+    pub fn build(&self) -> vector_common::Result<VrlDeserializer> {
+        let state = TypeState {
+            local: Default::default(),
+            external: ExternalEnv::default(),
+        };
+
+        match compile_vrl(
+            &self.vrl.source,
+            &vrl::stdlib::all(),
+            &state,
+            CompileConfig::default(),
+        ) {
+            Ok(result) => Ok(VrlDeserializer {
+                program: result.program,
+                timezone: self.vrl.timezone.unwrap_or(TimeZone::Local),
+            }),
+            Err(diagnostics) => Err(Formatter::new(&self.vrl.source, diagnostics)
+                .to_string()
+                .into()),
+        }
+    }
+
+    /// Return the type of event build by this deserializer.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema produced by the deserializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                schema::Definition::empty_legacy_namespace().unknown_fields(Kind::any())
+            }
+            LogNamespace::Vector => {
+                schema::Definition::new_with_default_metadata(Kind::any(), [log_namespace])
+            }
+        }
+    }
+}
+
+/// Deserializer that builds `Event`s from a byte frame containing logs compatible with VRL.
+#[derive(Debug, Clone)]
+pub struct VrlDeserializer {
+    program: Program,
+    timezone: TimeZone,
+}
+
+fn parse_bytes(bytes: Bytes, log_namespace: LogNamespace) -> Event {
+    let bytes_deserializer = BytesDeserializerConfig::new().build();
+    let log_event = bytes_deserializer.parse_single(bytes, log_namespace);
+    Event::from(log_event)
+}
+
+impl Deserializer for VrlDeserializer {
+    fn parse(
+        &self,
+        bytes: Bytes,
+        log_namespace: LogNamespace,
+    ) -> vector_common::Result<SmallVec<[Event; 1]>> {
+        let event = parse_bytes(bytes, log_namespace);
+        match self.run_vrl(event, log_namespace) {
+            Ok(event) => Ok(smallvec![event]),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl VrlDeserializer {
+    fn run_vrl(&self, event: Event, log_namespace: LogNamespace) -> vector_common::Result<Event> {
+        let mut runtime = Runtime::default();
+        let mut target = VrlTarget::new(event, self.program.info(), true);
+        match runtime.resolve(&mut target, &self.program, &self.timezone) {
+            Ok(_) => match target.into_events(log_namespace) {
+                TargetEvents::One(event) => Ok(event),
+                TargetEvents::Logs(_) => Err("multiple targets are not supported".into()),
```

Comment:
Why not? It seems reasonable to be able to generate multiple logs from a single input line 🤔

Reply:
It would require deep changes to this interface https://github.com/vectordotdev/vector/blob/v0.35.0/lib/codecs/src/decoding/mod.rs#L429 and I deemed it out of scope for this PR.

Reply:
Ah, gotcha 👍

Reply:
Actually, sorry, would it require any changes to that interface? It's already returning a `SmallVec`:

https://github.com/vectordotdev/vector/blob/e57c0c0e64bea6508ba0aae148d93525a4788669/lib/codecs/src/decoding/mod.rs#L429

Reply:
VRL actually already supports assigning an array to `.`, you can see how the `remap` transform handles this:

https://github.com/vectordotdev/vector/blob/56486bafe6ce41a7c92a11ccd0e2cf6e8f7ef838/src/transforms/remap.rs#L563-L569

Reply:
User-facing docs: https://vector.dev/highlights/2021-07-16-remap-multiple/

Reply:
I might be missing something but it seems like it'd be easy to support by just changing the signature of this `run_vrl` function to return a `SmallVec` of `Event`.

Reply:
This is pretty cool. I, will update soon.

Reply:
Done in https://github.com/vectordotdev/vector/pull/19825/commits/74849f1dd55f6571af92155263d07f76fe2a22fb.